pkg/spdk/replica.go

```diff
@@ -649,53 +689,376 @@ func (r *Replica) SnapshotDelete(spdkClient *spdkclient.Client, snapshotName str
 			childSvcLvol.Parent = parentSvcLvol.Name
-		}
-	} else {
-		if childSvcLvol != nil {
+		} else {
 			childSvcLvol.Parent = ""
 		}
 	}
+
+	delete(r.SnapshotMap, snapshotName)
 }
 
-func (r *Replica) removeLvolFromActiveChainWithoutLock(name string) int {
+func (r *Replica) removeLvolFromActiveChainWithoutLock(snapLvolName string) int {
 	pos := -1
 	for idx, lvol := range r.ActiveChain {
-		if lvol.Name == name {
+		if lvol.Name == snapLvolName {
 			pos = idx
 			break
 		}
 	}
 
+	prevChain := r.ActiveChain
 	if pos >= 0 && pos < r.ChainLength-1 {
-		r.ActiveChain = append([]*Lvol{}, r.ActiveChain[:pos]...)
-		r.ActiveChain = append(r.ActiveChain, r.ActiveChain[pos+1:]...)
+		r.ActiveChain = append([]*Lvol{}, prevChain[:pos]...)
+		r.ActiveChain = append(r.ActiveChain, prevChain[pos+1:]...)
 	}
 	r.ChainLength = len(r.ActiveChain)
 
 	return pos
 }
+
+func (r *Replica) RebuildingSrcStart(spdkClient *spdkclient.Client, localReplicaLvsNameMap map[string]string, dstReplicaName, dstRebuildingLvolAddress string) (err error) {
+	updateRequired := false
+
+	r.Lock()
+	defer func() {
+		r.Unlock()
+
+		if updateRequired {
+			r.UpdateCh <- nil
+		}
+	}()
+
+	if r.State != types.InstanceStateRunning {
+		return fmt.Errorf("invalid state %v for replica %s rebuilding src start", r.State, r.Name)
+	}
+	if r.rebuildingLvol != nil || r.rebuildingPort != 0 {
+		return fmt.Errorf("replica %s is being rebuilding hence it cannot be the source of rebuilding replica %s", r.Name, dstReplicaName)
+	}
+
+	dstRebuildingLvolIP, dstRebuildingLvolPort, err := net.SplitHostPort(dstRebuildingLvolAddress)
+	if err != nil {
+		return err
+	}
+	// TODO: After launching online rebuilding, the destination lvol name would be GetReplicaRebuildingLvolName(dstReplicaName)
+	dstRebuildingLvolName := dstReplicaName
+
+	if dstRebuildingLvolIP == r.IP {
+		dstReplicaLvsName := localReplicaLvsNameMap[dstReplicaName]
+		if dstReplicaLvsName == "" {
+			return fmt.Errorf("cannot find dst replica %s from the local replica map for replica %s rebuilding src start", dstReplicaName, r.Name)
+		}
+		r.rebuildingDstBdevName = spdktypes.GetLvolAlias(dstReplicaLvsName, dstRebuildingLvolName)
+	} else {
+		nvmeBdevNameList, err := spdkClient.BdevNvmeAttachController(dstRebuildingLvolName, helpertypes.GetNQN(dstRebuildingLvolName), dstRebuildingLvolIP, dstRebuildingLvolPort, spdktypes.NvmeTransportTypeTCP, spdktypes.NvmeAddressFamilyIPv4,
+			helpertypes.DefaultCtrlrLossTimeoutSec, helpertypes.DefaultReconnectDelaySec, helpertypes.DefaultFastIoFailTimeoutSec)
+		if err != nil {
+			return err
+		}
+		if len(nvmeBdevNameList) != 1 {
+			return fmt.Errorf("got zero or multiple results when attaching rebuilding dst lvol %s with address %s as a NVMe bdev: %+v", dstRebuildingLvolName, dstRebuildingLvolAddress, nvmeBdevNameList)
+		}
+		r.rebuildingDstBdevName = nvmeBdevNameList[0]
+	}
+
+	r.rebuildingDstReplicaName = dstReplicaName
+	r.rebuildingDstBdevType = spdktypes.BdevTypeNvme
```

Comment:
When will `rebuildingDstBdevType` be `spdktypes.BdevTypeLvol`? I thought it should be rebuilding happening on the same node, right? Just noticed you handle this case in RebuildingSrcFinish.

Reply:
Yes. When src replica and dst replica are on the same node, we cannot do nvme_controller_attach. Instead, we should directly use the local lvol bdev.